File: ZxSwift/Zx.swift
import Foundation

struct Zx {
  static let hLeftBorderSize = 48
  static let hScreenSize = 256
  static let hRightBorderSize = 48
  static let hBlankSize = 96
  
  static let vTopBorderSize = 48
  static let vScreenSize = 192
  static let vBottomBorderSize = 48
  static let vBlankSize = 24
  
  static let hLineSize = hLeftBorderSize + hScreenSize + hRightBorderSize
  static let vLineSize = vTopBorderSize + vScreenSize + vBottomBorderSize
  static let vMemSize = hLineSize * vLineSize
  
  static let hSize = hLineSize + hBlankSize
  static let vSize = vLineSize + vBlankSize
  static let vFrameSize = hSize * vSize
  
  static let barSize = vFrameSize / 28 - 43
  
  var vCounter: UInt16 = 0
  var hCounter: UInt16 = 0
  
  var barCounter: UInt16 = 0
  var barOn = false
  
  var memAddr : Int = 0
  
  var attr: UInt8 = 0
  var pixels: UInt8 = 0
  
  var flashCounter: UInt8 = 0
  
  var reg = Reg()
}

struct Reg {
  var pc: UInt16 = 0
  var sp: UInt16 = 0
  var lhs: UInt8 = 0
  var rhs: UInt8 = 0
  var a: UInt8 = 0
  var f: UInt8 = 0
  var a2: UInt8 = 0
  var f2: UInt8 = 0
  var wz: UInt16 = 0
  var bc: UInt16 = 0
  var de: UInt16 = 0
  var hl: UInt16 = 0
  var wz2: UInt16 = 0
  var bc2: UInt16 = 0
  var de2: UInt16 = 0
  var hl2: UInt16 = 0
  var ix: UInt16 = 0
  var iy: UInt16 = 0
  var i: UInt8 = 0
  var r: UInt8 = 0
  var op: UInt8 = 0
}

extension Zx {
  mutating func update(steps: Int, videoMem: UnsafeMutablePointer<UInt32>, romMem: UnsafePointer<UInt8>, scrMem: UnsafeMutablePointer<UInt8>) {
    for _ in 0..<steps {
      let hScreenWrite = hCounter >= Zx.hLeftBorderSize && hCounter < Zx.hLeftBorderSize &+ Zx.hScreenSize
      let vScreenWrite = vCounter >= Zx.vTopBorderSize && vCounter < Zx.vTopBorderSize &+ Zx.vScreenSize
      let screenWrite = hScreenWrite && vScreenWrite

      let hWrite = hCounter < Zx.hLineSize
      let vWrite = vCounter < Zx.vLineSize
      let memWrite = hWrite && vWrite

      if (memWrite) {
        let red, green, blue, bright: Bool
        if (screenWrite) {
          let bit = hCounter & 0x7
          if (bit == 0) {
            let hAddr = ((Int(hCounter) - Zx.hLeftBorderSize) >> 3) & 0x1f
            let vAddr = (Int(vCounter) - Zx.vTopBorderSize)
            let pixelsAddr = (((vAddr & 0xC0) | ((vAddr & 0x7) << 3) | ((vAddr & 0x38) >> 3)) << 5) | hAddr
            pixels = scrMem[pixelsAddr]
            let attrAddr = 0x1800 | ((vAddr >> 3) << 5) | hAddr
            attr = scrMem[attrAddr]
          }

          let pixelOn = (pixels & 0x80) != 0
          pixels = pixels << 1

          let flashOn = (attr & 0x80) != 0
          let alternateOn = flashCounter & 0x10 != 0
          
          let inkOn = flashOn && alternateOn ? !pixelOn : pixelOn
          red = attr & (inkOn ? 0x02 : 0x10) != 0
          green = attr & (inkOn ? 0x04 : 0x20) != 0
          blue = attr & (inkOn ? 0x01 : 0x08) != 0
          bright = (attr & 0x40) != 0
        } else {
          if (barOn) {
            red = true
            green = false
            blue = false
          } else {
            red = false
            green = true
            blue = true
          }
          bright = false
        }
        
        let component: UInt32 = bright ? 0xFF : 0xBB
        
        let col: UInt32 = 0xFF000000 | (blue ? component << 16 : 0) | (green ? component << 8 : 0) | (red ? component : 0)
        
        videoMem[memAddr] = col
        memAddr = memAddr &+ 1
      }

      hCounter = hCounter &+ 1
      if (hCounter == Zx.hSize) {
        hCounter = 0
        vCounter = vCounter &+ 1
        if (vCounter == Zx.vSize) {
          vCounter = 0
          memAddr = 0
          flashCounter = flashCounter &+ 1
        }
      }
      
      barCounter = barCounter &+ 1
      if (barCounter == Zx.barSize) {
        barCounter = 0
        barOn = !barOn
      }
    }
  }
}

File: ZxSwift/zx.c
#include "zx.h"

void zxUpdate(Zx* zxp, int steps) {
  struct Zx zx = *zxp;
  while (steps-- != 0) {
    int hScreenWrite = zx.hCounter >= hLeftBorderSize && zx.hCounter < hLeftBorderSize + hScreenSize;
    int vScreenWrite = zx.vCounter >= vTopBorderSize && zx.vCounter < vTopBorderSize + vScreenSize;
    int screenWrite = hScreenWrite && vScreenWrite;

    int hWrite = zx.hCounter < hLineSize;
    int vWrite = zx.vCounter < vLineSize;
    int memWrite = hWrite && vWrite;

    if (memWrite) {
      bool red, green, blue, bright;
      if (screenWrite) {
        int bit = zx.hCounter & 0x7;
        if (bit == 0) {
          int hAddr = ((zx.hCounter - hLeftBorderSize) >> 3) & 0x1f;
          int vAddr = zx.vCounter - vTopBorderSize;
          int pixelsAddr = (((vAddr & 0xC0) | ((vAddr & 0x7) << 3) | ((vAddr & 0x38) >> 3)) << 5) | hAddr;
          zx.pixels = zx.scrMem[pixelsAddr];
          int attrAddr = 0x1800 | ((vAddr >> 3) << 5) | hAddr;
          zx.attr = zx.scrMem[attrAddr];
        }

        bool pixelOn = (zx.pixels & 0x80) != 0;
        zx.pixels = zx.pixels << 1;

        bool flashOn = (zx.attr & 0x80) != 0;
        bool alternateOn = (zx.flashCounter & 0x10) != 0;

        bool inkOn = flashOn && alternateOn ? !pixelOn : pixelOn;
        red = (zx.attr & (inkOn ? 0x02 : 0x10)) != 0;
        green = (zx.attr & (inkOn ? 0x04 : 0x20)) != 0;
        blue = (zx.attr & (inkOn ? 0x01 : 0x08)) != 0;
        bright = (zx.attr & 0x40) != 0;
      } else {
        if (zx.barOn) {
          red = true;
          green = false;
          blue = false;
        } else {
          red = false;
          green = true;
          blue = true;
        }
        bright = false;
      }

      uint32_t component = bright ? 0xFF : 0xBB;

      uint32_t col = 0xFF000000 | (blue ? component << 16 : 0) | (green ? component << 8 : 0) | (red ? component : 0);

      zx.videoMem[zx.memAddr] = col;
      zx.memAddr = zx.memAddr + 1;
    }

    zx.hCounter++;
    if (zx.hCounter == hSize) {
      zx.hCounter = 0;
      zx.vCounter++;
      if (zx.vCounter == vSize) {
        zx.vCounter = 0;
        zx.memAddr = 0;
        zx.flashCounter++;
      }
    }

    zx.barCounter++;
    if (zx.barCounter == barSize) {
      zx.barCounter = 0;
      zx.barOn = !zx.barOn;
    }
  }
  *zxp = zx;
}
